### PR TITLE
Forward incoming SMS immediately

### PR DESCRIPTION
**Pull request by SCdF**
When an SMS is sent to the device and Gateway manages to wake up to deal with it, take the opportunity to get a request / response around from API. This should reduce latency between sending an SMS into our system and getting a response.

#138

### CODE REVIEW

**Comment by SCdF** [comment]
@garethbowen great feedback!

I added a note to the issue and also added the other end, attempting to send SMS back out (with tests), as well as creating an issue for 3 which I might tackle in the new year: https://github.com/medic/medic-gateway/issues/153

**Comment by dianabarsan** [comment]
hey @SCdF 


I'm AT-ing this on an Android 5 device and seeing this exception thrown every time Gateway receives an SMS and not seeing a corresponding request to API:

```
I/MedicGateway( 9929): IntentProcessor.onReceive() :: android.provider.Telephony.SMS_DELIVER
D/MedicGateway( 9929): medic.gateway.alert.SmsUdh :: from() :: pdu=[7, -111, 4, 39, 2, 80, 20, -7, 4, 11, -111, 4, 87, 53, 70, -106, -10, 0, 0, -111, 33, -111, 49, 18, 64, -128, 6, 78, 16, -68, 12, 47, 3]
D/MedicGateway( 9929): medic.gateway.alert.Db :: store() :: WtMessage@b898304a-0803-4923-b227-210287f7242b-WAITING
D/MedicGateway( 9929): medic.gateway.alert.ExternalLog :: log() :: WtMessage@b898304a-0803-4923-b227-210287f7242b-WAITING
D/MedicGateway( 9929): medic.gateway.alert.ExternalLog :: writeLine() :: json={"doc":{"sms_received":1576754466138,"sms_sent":1576754464000,"content":"N pepe","from":"*********","id":"b898304a-0803-4923-b227-210287f7242b"},"type":"wt_message"}
D/MedicGateway( 9929): medic.gateway.alert.Db :: getWtMessages() :: item fetch count: 1
D/MedicGateway( 9929): medic.gateway.alert.Db :: getWoMessageStatusUpdates() :: item fetch count: 0
D/MedicGateway( 9929): medic.gateway.alert.SettingsStore :: loading for context: android.app.ReceiverRestrictedContext@b9aab27
D/MedicGateway( 9929): medic.gateway.alert.Settings :: Settings() webappUrl=https://gateway:****@89108a2f.eu.ngrok.io/api/sms
I/MedicGateway( 9929): Polling webapp (forwarding 1 messages & 0 status updates)...
D/MedicGateway( 9929): SimpleJsonClient2.post() :: url=https://gateway:****@89108a2f.eu.ngrok.io/api/sms
D/MedicGateway( 9929): SimpleJsonClient2.post() :: url=https://gateway:****@89108a2f.eu.ngrok.io/api/sms
------- EXCEPTION HERE ------ (adding comment for visibility)
I/MedicGateway( 9929): Exception caught trying to poll webapp: null :: android.os.NetworkOnMainThreadException
I/MedicGateway( 9929): android.os.NetworkOnMainThreadException
I/MedicGateway( 9929): 	at android.os.StrictMode$AndroidBlockGuardPolicy.onNetwork(StrictMode.java:1155)
I/MedicGateway( 9929): 	at com.android.org.conscrypt.OpenSSLSocketImpl$SSLInputStream.read(OpenSSLSocketImpl.java:657)
I/MedicGateway( 9929): 	at com.android.okio.Okio$2.read(Okio.java:113)
I/MedicGateway( 9929): 	at com.android.okio.RealBufferedSource.exhausted(RealBufferedSource.java:58)
I/MedicGateway( 9929): 	at com.android.okhttp.internal.http.HttpConnection.isReadable(HttpConnection.java:140)
I/MedicGateway( 9929): 	at com.android.okhttp.Connection.isReadable(Connection.java:255)
I/MedicGateway( 9929): 	at com.android.okhttp.internal.http.RouteSelector.next(RouteSelector.java:110)
I/MedicGateway( 9929): 	at com.android.okhttp.internal.http.HttpEngine.connect(HttpEngine.java:272)
I/MedicGateway( 9929): 	at com.android.okhttp.internal.http.HttpEngine.sendRequest(HttpEngine.java:211)
I/MedicGateway( 9929): 	at com.android.okhttp.internal.http.HttpURLConnectionImpl.execute(HttpURLConnectionImpl.java:373)
I/MedicGateway( 9929): 	at com.android.okhttp.internal.http.HttpURLConnectionImpl.connect(HttpURLConnectionImpl.java:106)
I/MedicGateway( 9929): 	at com.android.okhttp.internal.http.HttpURLConnectionImpl.getOutputStream(HttpURLConnectionImpl.java:208)
I/MedicGateway( 9929): 	at com.android.okhttp.internal.http.DelegatingHttpsURLConnection.getOutputStream(DelegatingHttpsURLConnection.java:218)
I/MedicGateway( 9929): 	at com.android.okhttp.internal.http.HttpsURLConnectionImpl.getOutputStream(HttpsURLConnectionImpl.java:25)
I/MedicGateway( 9929): 	at medic.gateway.alert.SimpleJsonClient2.post(SimpleJsonClient2.java:80)
I/MedicGateway( 9929): 	at medic.gateway.alert.SimpleJsonClient2.post(SimpleJsonClient2.java:64)
I/MedicGateway( 9929): 	at medic.gateway.alert.WebappPoller.pollWebapp(WebappPoller.java:51)
I/MedicGateway( 9929): 	at medic.gateway.alert.IntentProcessor.handleSmsReceived(IntentProcessor.java:77)
I/MedicGateway( 9929): 	at medic.gateway.alert.IntentProcessor.onReceive(IntentProcessor.java:45)
I/MedicGateway( 9929): 	at android.app.ActivityThread.handleReceiver(ActivityThread.java:2712)
I/MedicGateway( 9929): 	at android.app.ActivityThread.access$1700(ActivityThread.java:144)
I/MedicGateway( 9929): 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1449)
I/MedicGateway( 9929): 	at android.os.Handler.dispatchMessage(Handler.java:102)
I/MedicGateway( 9929): 	at android.os.Looper.loop(Looper.java:155)
I/MedicGateway( 9929): 	at android.app.ActivityThread.main(ActivityThread.java:5696)
I/MedicGateway( 9929): 	at java.lang.reflect.Method.invoke(Native Method)
I/MedicGateway( 9929): 	at java.lang.reflect.Method.invoke(Method.java:372)
I/MedicGateway( 9929): 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1028)
I/MedicGateway( 9929): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:823)
```

**Comment by SCdF** [comment]
@dianabarsan thanks for the catch! Surprising int. tests didn't catch this honestly, seems like a pretty straight forward mistake where me not being an android dev meant I messed up something pretty basic.

I'll look into it.

**Comment by SCdF** [comment]
@dianabarsan I fixed this by moving the IO out of the main thread. The test is gross. Let me know _how_ gross ;-)

**Comment by dianabarsan** [comment]
@SCdF me and Java are parallel, I'm really not the right person for code review :( 
I can try to understand it, but in terms of reviewing code quality, I am be completely useless. 
I'm happy to AT though!

**Comment by SCdF** [comment]
No worries, @garethbowen can you take a look?

File: src/main/java/medic/gateway/alert/IntentProcessor.java [hunk 1]
```diff
@@ -72,6 +73,23 @@ private void handleSmsReceived(Context ctx, Intent intent) {
...
+		// there is no way to delete the message.
+		abortBroadcast();
+	}
+}
+
+class AsyncPoller extends AsyncTask<Context, Void, Void> {
+	@Override
+	protected Void doInBackground(Context... contexts) {
+		Context ctx = contexts[0];
+		logEvent(ctx, "This code ran yo");
```
Comment (garethbowen) [inline src/main/java/medic/gateway/alert/IntentProcessor.java]: Yeah, nah...

File: src/main/java/medic/gateway/alert/IntentProcessor.java [hunk 2]
```diff
@@ -81,6 +85,37 @@ private void handleSmsReceived(Context ctx, Intent intent) {
 	}
 }
 
+class AsyncPoller extends AsyncTask<Context, Void, Void> {
+	@Override
+	protected Void doInBackground(Context... contexts) {
+		Context ctx = contexts[0];
+		logEvent(ctx, "This code ran yo");
```
Comment (SCdF) [inline src/main/java/medic/gateway/alert/IntentProcessor.java]:
```suggestion
```

**Pull request merged**